File: backend/src/routes/referrals.js
const express = require('express');
const prisma = require('../config/db');
const { authenticate } = require('../middleware/auth');

const router = express.Router();
router.use(authenticate);

// ─── Get My Referral Code & Stats ────────────────────
router.get('/my-code', async (req, res, next) => {
  try {
    const user = await prisma.user.findUnique({
      where: { id: req.user.id },
      select: { referralCode: true },
    });

    const referralCount = await prisma.referral.count({
      where: { referrerId: req.user.id },
    });

    res.json({
      referralCode: user.referralCode,
      referralCount,
      referralLink: `${process.env.FRONTEND_URL}/register?ref=${user.referralCode || ''}`,
    });
  } catch (error) {
    next(error);
  }
});

module.exports = router;

File: backend-go/internal/handlers/referrals.go
package handlers

import (
	"fmt"
	"net/http"
)

// ─── Get My Referral Code & Stats ────────────────────

func (h *Handler) GetMyReferralCode(w http.ResponseWriter, r *http.Request) {
	user := GetUser(r)
	if user == nil {
		RespondError(w, http.StatusUnauthorized, "Not authenticated.")
		return
	}

	ctx := r.Context()

	// Get user's referral code
	var referralCode *string
	err := h.DB.QueryRow(ctx,
		`SELECT referral_code FROM users WHERE id = $1`, user.ID,
	).Scan(&referralCode)
	if err != nil {
		RespondError(w, http.StatusInternalServerError, "Failed to fetch referral code.")
		return
	}

	// Count referrals where referrer_id = user.id
	var referralCount int
	err = h.DB.QueryRow(ctx,
		`SELECT COUNT(*) FROM referrals WHERE referrer_id = $1`, user.ID,
	).Scan(&referralCount)
	if err != nil {
		referralCount = 0
	}

	code := ""
	if referralCode != nil {
		code = *referralCode
	}

	RespondJSON(w, http.StatusOK, map[string]interface{}{
		"referralCode":  code,
		"referralCount": referralCount,
		"referralLink":  fmt.Sprintf("%s/register?ref=%s", h.Cfg.FrontendURL, code),
	})
}
